### ISSUE TITLE
[Bug] multiple `sgl.Runtime` instances compete for port 10000

### ISSUE THREAD

**Issue by mantle2048** [sgl-project/sglang#2430]
### Checklist

- [X] 1. I have searched related issues but cannot get the expected help.
- [X] 2. The bug has not been fixed in the latest version.
- [X] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [ ] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [X] 5. Please use English, otherwise it will be closed.

### Describe the bug

Hi,

I am working on offline inference for an image_qa task, following the code example provided in [examples/frontend_language/quick_start/local_example_llava_next.py](https://github.com/sgl-project/sglang/blob/main/examples/frontend_language/quick_start/local_example_llava_next.py).

Since I need to evaluate multiple models, I plan to run one `sgl.Runtime` on each GPU and specify a different port for each instance.

However, when I set different ports and execute the file multiple times via the command line to launch several Runtime instances (with a few seconds interval between launches), I get an error.

```bash
[2024-12-10 15:50:02] ERROR:    [Errno 98] error while attempting to bind on address ('127.0.0.1', 10000): address already in use
```

After checking the code, I think the issue may lie in the for-loop mentioned below.

```python
# https://github.com/sgl-project/sglang/blob/main/python/sglang/srt/server.py#L906-L909
# Pre-allocate ports
for port in range(10000, 40000):
    if is_port_available(port):
        break
        port += 1
self.server_args.port = port
```

Even if the port is explicitly specified in `**kwargs` , sglang still executes this for-loop to select a port. This leads to a scenario where, if the previously launched Runtime has not yet occupied port 10000, the later-launched instance mistakenly selects port 10000.

After commenting this for-loop, everything works as expected. 

I'm wondering if there's a more reasonable way to handle port configuration for Runtime. For instance, if the port parameter is provided via the `*args` or `**kwargs` in `__init__`, could the for-loop be skipped altogether.

### Reproduction

model: qwen2-vl-instruct 7b
command: 

```bash
# in tmux terminal 0
CUDA_VISIBLE_DEVICES=0 examples/frontend_language/quick_start/local_example_llava_next.py
# in tmux terminal 1
CUDA_VISIBLE_DEVICES=1 examples/frontend_language/quick_start/local_example_llava_next.py
```

### Environment

Python: 3.11.10 (main, Oct  3 2024, 07:29:13) [GCC 11.2.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA A800-SXM4-80GB
GPU 0,1,2,3,4,5,6,7 Compute Capability: 8.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.1, V12.1.105
CUDA Driver Version: 535.54.03
PyTorch: 2.5.1+cu121
sglang: 0.3.6.post2
flashinfer: 0.1.6+cu121torch2.4
triton: 3.1.0
transformers: 4.46.2
torchao: 0.6.1
numpy: 1.26.3
aiohttp: 3.10.10
fastapi: 0.115.4
hf_transfer: 0.1.8
huggingface_hub: 0.26.2
interegular: 0.3.3
modelscope: 1.20.1
orjson: 3.10.11
packaging: 24.1
psutil: 6.1.0
pydantic: 2.9.2
multipart: 0.0.17
zmq: 26.2.0
uvicorn: 0.32.0
uvloop: 0.21.0
vllm: 0.6.4.post1
openai: 1.57.0
anthropic: 0.40.0
decord: 0.6.0
NVIDIA Topology: 
	[4mGPU0	GPU1	GPU2	GPU3	GPU4	GPU5	GPU6	GPU7	NIC0	NIC1	NIC2	NIC3	CPU Affinity	NUMA Affinity	GPU NUMA ID[0m
GPU0	 X 	NV8	NV8	NV8	NV8	NV8	NV8	NV8	PHB	PHB	PHB	PHB	0-95		N/A		N/A
GPU1	NV8	 X 	NV8	NV8	NV8	NV8	NV8	NV8	PHB	PHB	PHB	PHB	0-95		N/A		N/A
GPU2	NV8	NV8	 X 	NV8	NV8	NV8	NV8	NV8	PHB	PHB	PHB	PHB	0-95		N/A		N/A
GPU3	NV8	NV8	NV8	 X 	NV8	NV8	NV8	NV8	PHB	PHB	PHB	PHB	0-95		N/A		N/A
GPU4	NV8	NV8	NV8	NV8	 X 	NV8	NV8	NV8	PHB	PHB	PHB	PHB	0-95		N/A		N/A
GPU5	NV8	NV8	NV8	NV8	NV8	 X 	NV8	NV8	PHB	PHB	PHB	PHB	0-95		N/A		N/A
GPU6	NV8	NV8	NV8	NV8	NV8	NV8	 X 	NV8	PHB	PHB	PHB	PHB	0-95		N/A		N/A
GPU7	NV8	NV8	NV8	NV8	NV8	NV8	NV8	 X 	PHB	PHB	PHB	PHB	0-95		N/A		N/A
NIC0	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	 X 	PHB	PHB	PHB				
NIC1	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	 X 	PHB	PHB				
NIC2	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	 X 	PHB				
NIC3	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	PHB	 X 				

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3

ulimit soft: 1024

**Comment by ByronHsu** [comment]
I think this is a bug. @merrymercy do you know why we overwrite the port from serving_args with the one from the for loop?

**Comment by merrymercy** [comment]
Yes. This is a bug. @mantle2048 Can you help us fix it? If the port is provided, we can skip the for loop or do something like
```
        for port in range(given_port, 40000):
            if is_port_available(port):
                break
            port += 1
```
